Make classes in diffgeom module immutable

mcpl-sympy opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #19321
Fixes #19318

#### Brief description of what is fixed or changed

`Manifold.patches` attribute is deprecated.
`Patch.coord_systems` attribute is deprecated.
Class signature *names* of `CoordSystem` is deprecated. `CoordinateSymbol` class and `CoordSystem.symbols` attribute are introduced instead.
`CoordSystem.transforms` attribute,  `CoordSystem.connect_to` method, and `CoordSystem.coord_tuple_transform_to` method are deprecated. Class signature *relations*, `CoordSystem.transformation` and `CoordSystem.transform` methods are introduced instead.
`CoordSystem.jacobian_determinant` method is introduced.
 methods are introduced instead.

#### Other comments

`Manifold.patches`, `Patch.coord_systems`, and `CoordSystem.transforms` attributes are deprecated because these make classes mutable.  
`CoordSystem._dummies` is replaced with `CoordinateSymbol` class to use `Dummy` less. Also, this makes the coordinate symbols know which coordinate system they belong to.  

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- diffgeom
  - `Manifold.patches` attribute is deprecated.
  - `Patch.coord_systems` attribute is deprecated.
  - Class signature *names* of `CoordSystem` is deprecated. `CoordinateSymbol` class and `CoordSystem.symbols` attribute are introduced instead.
  - `CoordSystem.transforms`,  `CoordSystem.connect_to`, and `CoordSystem.coord_tuple_transform_to` are deprecated.  Class signature *relations*, `CoordSystem.transformation` and `CoordSystem.transform` are introduced instead.
  - `CoordSystem.jacobian_determinant` method is introduced.

<!-- END RELEASE NOTES -->

mcpl-sympy commented:
@sylee957 Please review this.

sylee957 commented:
Regarding the deprecation issues, I think that `Manifold.patches` and `Patch.coord_systems` can raise some deprecation warnings by using some custom objects that wraps on `list` and `dict` that raises some warnings on item access or iterations.
But if the original mutable designs have other critical issues like memory leak, this can be ignored.

mcpl-sympy commented:
@sylee957  
Sounds alright. I'll try that one.

sylee957 commented on sympy/diffgeom/rn.py:
```diff
@@ -50,40 +54,34 @@
+R3_c = CoordSystem(
+    'cylindrical', R3_origin, ['rho', 'psi', 'z'],
+    {
+        R3_r: (
```
Does `R3_s` get linked automatically here?
One of the problem of this immutable design is that the you cannot link `R3_s` before defining.

I have an another idea as making the `transforms` dict totally external to the class, and supplied as function arguments when needed, and encourage users to build the charts outside of classes.

mcpl-sympy replied:
In my design, `R3_s` never exists in `R3_c` nor in `R3_r`. Relation between two systems never exists in earlier-generated system. It is stored in later-generated one.  
Therefore, in this case, `R3_r` contains no relation. `R3_c` contains bidirectional relation with `R3_r`. `R3_s` contains bidirectional relations with `R3_r` and `R3_c`.

I choose this hierarchial design to include 'transform_dict` in `srepr` printing and `_hashable_content` of `CoordSystem`.  
1. `_hashable_content` needs to contain the relation. It's because two coordinate system which have same relations with reference systems are efficiently same. In other words, even if two systems have same name, patch, and notation, they are different if they have different relations with reference system.  
2. It is SymPy's rule that `eval(srepr(obj))` returns mathematically same object. If we don't include `transform_dict` in `srepr` printing, the result of `eval(srepr(_))` will not contain any transform relation, making itself completely useless. And, in order to contain `transform_dict` in `srepr` printing, it should be immutable.

I'd like to hear your thoughts on this two issues. Also, if you have idea to circumvent these, please let me know. I'm all ears!

sylee957 replied:
I see the coordinate transformation can work even if the chart is defined in only one class
But I'd suggest them to have have more symmetric structures.

Although this can be more difficult to resolve, however, I think that there can be something like `CoordSystemSymbol` defined prior to creating `CoordSystem` and the dictionary can made symmetric by taking reference to such symbols.

sylee957 replied:
My hunch of this is replacing the name of `CoordSystem` to some symbol, and tie the transformations with the name rather than `CoordSystem`
I think that the transformations dict need not even be defined bidirectional if this can be done.

```diff
diff --git a/sympy/diffgeom/diffgeom.py b/sympy/diffgeom/diffgeom.py
index 97f4cf1c34..d7af6f0567 100644
--- a/sympy/diffgeom/diffgeom.py
+++ b/sympy/diffgeom/diffgeom.py
@@ -344,11 +344,10 @@ def coord_tuple_transform_to(self, to_sys, coords):

         See the docstring of ``CoordSystem`` for examples."""
         if self != to_sys:
-
-            if to_sys in self.transform_dict:
-                transf = self.transform_dict[to_sys][0]
-            elif self in to_sys.transform_dict:
-                transf = to_sys.transform_dict[self][1]
+            if to_sys.name in self.transform_dict:
+                transf = self.transform_dict[to_sys.name][0]
+            elif self.name in to_sys.transform_dict:
+                transf = to_sys.transform_dict[self.name][1]
             else:
                 transf = self.transforms[to_sys]
             coords = transf(*coords)
diff --git a/sympy/diffgeom/rn.py b/sympy/diffgeom/rn.py
index c07605ac6b..61504457c4 100644
--- a/sympy/diffgeom/rn.py
+++ b/sympy/diffgeom/rn.py
@@ -25,11 +25,23 @@
 R2 = Manifold('R^2', 2)  # type: Any
 # Patch and coordinate systems.
 R2_origin = Patch('origin', R2)  # type: Any
-R2_r = CoordSystem('rectangular', R2_origin, ['x', 'y'])  # type: Any
+
+R2_r_sym = Dummy('R2_r_sym')
+R2_p_sym = Dummy('R2_p_sym')
+
+R2_r = CoordSystem(
+    R2_r_sym, R2_origin, ['x', 'y'],
+    {
+        R2_p_sym: (
+            Lambda((x, y), [sqrt(x**2 + y**2), atan2(y, x)]),
+            Lambda((r, theta), [r*cos(theta), r*sin(theta)])
+        )
+    }
+    )  # type: Any
 R2_p = CoordSystem(
-    'polar', R2_origin, ['r', 'theta'],
+    R2_p_sym, R2_origin, ['r', 'theta'],
     {
-        R2_r: (
+        R2_r_sym: (
             Lambda((r, theta), [r*cos(theta), r*sin(theta)]),
             Lambda((x, y), [sqrt(x**2 + y**2), atan2(y, x)])
         )
@@ -59,24 +71,45 @@
 R3 = Manifold('R^3', 3)  # type: Any
 # Patch and coordinate systems.
 R3_origin = Patch('origin', R3)  # type: Any
-R3_r = CoordSystem('rectangular', R3_origin, ['x', 'y', 'z'])  # type: Any
+
+R3_r_sym = Dummy('R3_r_sym')
+R3_c_sym = Dummy('R3_c_sym')
+R3_s_sym = Dummy('R3_s_sym')
+
+R3_r = CoordSystem(
+    R3_r_sym, R3_origin, ['x', 'y', 'z'],
+    {
+        R3_c_sym: (
+            Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z]),
+            Lambda((rho, psi, z), [rho*cos(psi), rho*sin(psi), z])
+        ),
+        R3_s_sym: (
+            Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)]),
+            Lambda((r, theta, phi), [r*sin(theta)*cos(phi), r*sin(theta)*sin(phi), r*cos(theta)]),
+        )
+    }
+)  # type: Any
 R3_c = CoordSystem(
-    'cylindrical', R3_origin, ['rho', 'psi', 'z'],
+    R3_c_sym, R3_origin, ['rho', 'psi', 'z'],
     {
-        R3_r: (
+        R3_r_sym: (
             Lambda((rho, psi, z), [rho*cos(psi), rho*sin(psi), z]),
             Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z])
+        ),
+        R3_s_sym: (
+            Lambda((rho, psi, z), [sqrt(rho**2 + z**2), acos(z/sqrt(rho**2 + z**2)), psi]),
+            Lambda((r, theta, phi), [r*sin(theta), phi, r*cos(theta)])
         )
     }
 )  # type: Any
 R3_s = CoordSystem(
-    'spherical', R3_origin, ['r', 'theta', 'phi'],
+    R3_s_sym, R3_origin, ['r', 'theta', 'phi'],
     {
-        R3_r: (
+        R3_r_sym: (
             Lambda((r, theta, phi), [r*sin(theta)*cos(phi), r*sin(theta)*sin(phi), r*cos(theta)]),
             Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)])
         ),
-        R3_c: (
+        R3_c_sym: (
             Lambda((r, theta, phi), [r*sin(theta), phi, r*cos(theta)]),
             Lambda((rho, psi, z), [sqrt(rho**2 + z**2), acos(z/sqrt(rho**2 + z**2)), psi])
         )
(END)
```

sylee957 replied:
Maybe it can even be better to define the names as strings and make the dictionary keys as strings.
no one may define the coordinate systems as same strings.

mcpl-sympy replied:
@sylee957 
Coincidentally, your suggestion is exactly same with my previous attempt. I ditched that attempt because of its innate restriction:

When you define 'R2_r', you included its relation with 'R2_p'. Then, how can we define a third coordinate system and connect it to 'R2_r' without mutating it?  
My answer was we cannot. This made the entire system locked, non-expandable; which was a serious problem.

mcpl-sympy replied:
In my case, expandable coordinate system is necessary in field of fluid mechanics. When analyzing the confined flow, one may need to define 'stretched coordinates', connecting it to existing ordinary coordinate systems.

Drawback of making the systems locked exceeds the drawback of non-symmetric relations, I reckon.

sylee957 replied:
I don't think that this disallows defining coordinate systems in non symmetric way.
If the name identifier is in either of the dictionary, it will work like your previous implementation.

mcpl-sympy replied:
@sylee957  
Oh, I see. It will be fine if we allow asymmetric dictionaries.  
But then, wouldn't it be unnecessary to keep symmetric bi-directional dictionary if we can just make it asymmetric? It's only `transform_dict` which is asymmetric; every other features such as `coord_tuple_transform_to` are not.

mcpl-sympy replied:
Maybe I can make it like this:

1. `R2_r` and `R2_p` know each other symmetrically and unidirectionally.  
2. If new coordinate system is introduced, it is connected to `R2_r` and `R2_p` asymmetrically and bidirectionally.

mcpl-sympy replied:
About using string as dictionary keys: this will be problematic if one want to scrap existing system and define another with same name. 

Let's think that one imported `R2_r` from `diffgeom.rn` but want to define a new one with name `'polar'`. Since importing `R2_r` runs `diffgeom.rn` and registers `'polar'` to it, this will raise some problems. I think `Dummy` will be more suitable for the key.

sylee957 replied:
For default classes, what about like 
```
R3_r = CoordSystem(
    R3_r_sym, R3_origin, ['x', 'y', 'z'],
    {
        R3_c_sym: (Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z]),),
        R3_s_sym: (Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)]),)
    }
)
```
so the transformations can be defined symmetrically, but without redundant edges for inverse transformations?

mcpl-sympy replied:
@sylee957 
Yep, I thought the same.  
By the way, although I'm adopting your suggestion, I'm not sure about the advantage we'll have by making the dictionary symmetric. Could you give a lead, perhaps?

sylee957 replied:
I don't think that making the composition non symmetric is a significant issue, but I think that it was a subset of the problems of restricting a lot of freedom of how the composition graph can be formed.

However, for resolving this issue, I don't think that my above suggestion can suffice, but making the transformation dictionary not stored in the object signature could be a more valid idea.

sylee957 replied:
I have my own branch of making this possible
https://github.com/sylee957/sympy/tree/diffgeom_transform

```
R2_r = CoordSystem('rectangular', R2_origin, ['x', 'y'])
R2_p = CoordSystem('polar', R2_origin, ['r', 'theta'])
R3_r = CoordSystem('rectangular', R3_origin, ['x', 'y', 'z'])
R3_c = CoordSystem('cylindrical', R3_origin, ['rho', 'psi', 'z'],)
R3_s = CoordSystem('spherical', R3_origin, ['r', 'theta', 'phi'],)
```

And I think that it is possible to separate the transformation entirely from the coordinate systems, so they are can be manipulated independently,

However, this makes it bulkier that it should be invoked by keyword arguments
`polar.coord_tuple_transform_to(rect, [0, 2], transforms=transforms)`
But I think we can rather have a full symbolic object specialized for coordinate transformations and dispatch stuff like `coord_tuple_transform_to` and rcalls to there.

I don't think that it was hierarchically good that the objects store an another object that should be operating on top of this, as well as mutability.

mcpl-sympy replied:
@sylee957 
Does this design allow pickling?

mcpl-sympy replied:
@sylee957 
Maybe we can define a `dict` instance in `diffgeom` and make the other methods to use that dictionary as default. Then, as long as the user stores all transform relation in that instance, `polar.coord_tuple_transform_to(rect, [0, 2])` will be suffice for most cases.

mcpl-sympy replied:
When will you make your pull request? Shall I merge your work into this PR?

mcpl-sympy commented:
@sylee957

I added deprecation wrapper for list and dict. Please review this

sylee957 commented:
I also think that the future about #19318 should be addressed here because `fill_in_gaps` will not be valid after `connect_to` is deprecated.
I'd have an idea as implementing that as a external function that takes multiple `dict` instance and can figure out a composition, but I'm not sure if you may say that this is not sufficient for your application.

mcpl-sympy commented:
@sylee957 
Yes, I was planning for that, either.  
My idea was like this: If we have explicit relation between A-B and B-C, then A-C are implicitly related as well. It's just that deriving the relation will be time-consuming because `solve` will be involved.  
So, instead of developing `fill_in_gaps`, I was going to make a method which solves and returns the relation between A and C then caches it. (Perhaps by using `cachit` decorator).

mcpl-sympy commented:
I originally wrote this comment [here](https://github.com/mcpl-sympy/sympy/pull/2), but I'll quote it in this PR so that other people can see and share their opinions.

> I'm sorry, but I've been thinking of this awhile and I have to confess; I'm still not convinced that transform relation should be kept external.
> 
> For [curvilinear coordinates](https://en.wikipedia.org/wiki/Curvilinear_coordinates), relation with its 'parent' Cartesian coordinate system is the most important attribute; it literally defines the very nature of the system. Not only its Jacobian matrix, but its differential operator, metric tensor, everything depends on this relation.
> 
> I just can't convince myself that a single coordinate system is indetermined and can be Cartesian, curvilinear, affine, or anything else depending on the external dictionary. For me, this is tantamount to making the `args` of `Add`, `Mul`, or any other class undetermined and letting it dependent to external container - making it effectively mutable.
> 
> If we want to make a directed graph between the classes, we might as well make an algorithm to 'crawl' every connected system by querying whether one system 'knows' the other. If this is unacceptable because of performance reason, then it would be better to just make `CoordSystem` the subclass of `object` and let it be freely mutable.

sylee957 commented:
> For curvilinear coordinates, relation with its 'parent' Cartesian coordinate system is the most important attribute; it literally defines the very nature of the system.

I don't get what is such 'nature'. We can think as a whole collection of `(CoordSys1, CoordSys2, ..., Transforms)` to be definition of the system.

> Not only its Jacobian matrix, but its differential operator, metric tensor, everything depends on this relation.

As far as I see how `diffgeom` offers for now, all other stuff are derived from the jacobian matrix. I'm not sure that if are some more advanced concepts that cannot be derived from it, But again I believe that anything can be wrapped up in an another object that defines the concept, if the stuff have to be mutually defined.

> For me, this is tantamount to making the args of Add, Mul, or any other class undetermined and letting it dependent to external container

Addition and multiplication is not well defined from the context as elements of sets, but from the context of rings or fields. 
And it is possible to let such algebra as an optional context.

> making it effectively mutable.

I don't think that it should be sympy's scope to have some incremental registry because I have seen that it had been becoming problems.
But I left is as user's choice to use the mutable context or not. If you just pass on mutable dictionary to `transforms` keyword argument, it would suffice.

> If we want to make a directed graph between the classes, we might as well make an algorithm to 'crawl' every connected system by querying whether one system 'knows' the other. 

Why not crawl the graph defined in transformation dict?

> then it would be better to just make CoordSystem the subclass of object and let it be freely mutable.

I believe that my approach supersets any designs you suggest. Basically, if you pass some part of dictionary, it should be equivalent to wrapping the dictionary with object, and if you pass some mutable dictionary, it should be equivalent to defining mutable object.

mcpl-sympy commented:
@sylee957  
I appreciate your comment. Thanks to it, I have a new perspective for this matter now.

Maintaining your design, I've come up with an idea of introducing `CoordSystems`class (its name may change), which takes coordinate systems and transformation dictionary as arguments.

I'll work on this and show you when it's done. Thanks again!

mcpl-sympy commented:
I've merged your PR, but please don't merge this yet.

mcpl-sympy commented:
@sylee957 
OK, it's done. Some changes have been made after your work, though.

1. I introduced `CoordSystemStructure` class. This is a subclass of `Dict`, but its `__getitem__` method is overrided so that inverse transform can be derived. After this PR, issue #19318 will be dealt with by adding additional method to this class.

2. Rather than including `CoordSystemStructure` in method signature, I added it in class signatures of `Point`, `BaseScalarField`, etc. With this, these classes now represent "geometric object defined with certain context". This structure exists in `args`, so one can introduce another coordinate system - by defining new `CoordSystemStructure` and replacing it into geometric entities.

Please review this!

mcpl-sympy commented:
@sylee957, I added a feature to resolve #19318. I will be much obliged if you kindly review this.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -130,110 +151,34 @@ class CoordSystem(Atom):
+            "1.7",
+            {}
+        ) # deprecated
         obj._dummies = [Dummy(str(n)) for n in names]
         obj._dummy = Dummy()
```
I was thinking of removing these dummies.
Have you tried anything successful?

mcpl-sympy replied:
Yes. In fact, I have had a plan of removing these after this PR, as well. I'll make another PR and ping you there.

mcpl-sympy replied:
`_dummy` is used in `BaseVectorField` in some strange way as an attribute of 'dummy function'. In my local repository, I have `DummyFunction` which is subclass of `Function`. I am planning to contribute that to SymPy, and after then we can deal with this issue.

Until then, removing these are out of this PR's scope so I'll mark this conversation as resolved.

sylee957 commented:
I notice a lot of slowdowns in `test_function_diffgeom_book` when I set the environment variable `SYMPY_USE_CACHE=no`,

mcpl-sympy commented:
@sylee957 
I tried some past commits, and here are the times took in `test_function_diffgeom_book` with cache off.

|HEAD|time (s)|
|---|---|
|master|9.7|
|87e49fd|9.3|
|1783523|12.6|
|c24965d (current)| 13.6|

Any suggestion, please?

sylee957 commented:
master, cache - 1.17 seconds
master, nocache - 11.09 seconds
head, cache - 1.74 seconds
head, nocache - 15.79 seconds

Okay, it could already been 10 times slower in the master. But I want to figure out which whether any changes here are introducing some slowdowns.

mcpl-sympy commented:
@sylee957 
The first slowdown is encountered in 1783523, so I would like to start the investigation there.

mcpl-sympy commented:
In master, each test took

|function|time (s)|
|---|---|
|test_functional_diffgeom_ch2|0.26|
|test_functional_diffgeom_ch3|1.53|
|test_functional_diffgeom_ch4|4.73|
|test_functional_diffgeom_ch6|0.26|


In 1783523, each test took

|function|time (s)|
|---|---|
|test_functional_diffgeom_ch2|0.23|
|test_functional_diffgeom_ch3|2.08|
|test_functional_diffgeom_ch4|**6.37**|
|test_functional_diffgeom_ch6|0.33|

In HEAD, each test took

|function|time (s)|
|---|---|
|test_functional_diffgeom_ch2|0.24|
|test_functional_diffgeom_ch3|**2.17**|
|test_functional_diffgeom_ch4|**6.57**|
|test_functional_diffgeom_ch6|**1.41**|

mcpl-sympy commented:
`df(R2.e_x)`, `df(R2.e_y)`, `R2.dr(circ)`, `R2.dtheta(circ)` were the most time-consuming in `test_functional_diffgeom_ch4`.

sylee957 commented:
I only recognize that some behavior of substitution changed in `CoordSystem.jacobian`. And it looks like `subs` is doing more work than before.
But I'm not sure this alone had made things slower.

sylee957 commented:
```diff
$ git diff
diff --git a/sympy/diffgeom/diffgeom.py b/sympy/diffgeom/diffgeom.py
index aca1d1322c..51c82b828b 100644
--- a/sympy/diffgeom/diffgeom.py
+++ b/sympy/diffgeom/diffgeom.py
@@ -750,7 +750,7 @@ def __call__(self, scalar_field):
         d_result = scalar_field.subs(list(zip(base_scalars, d_funcs)))
         d_result = d_result.diff(d_var)
         # Second step: e_x(x) -> 1 and e_x(r) -> cos(atan2(x, y))
-        coords = [Dummy() for _ in range(self._coord_sys.dim)]
+        coords = self._coord_sys._dummies
         d_funcs_deriv = [f.diff(d_var) for f in d_funcs]
         d_funcs_deriv_sub = []
         for b in base_scalars:
```

I guess this just made things slower because it doesn't equal the local dummies used by `jacobian`, and subs is expensive operation.
So I'd dump the attempt to replace `_dummies` cache for now.

mcpl-sympy commented:
@sylee957 Indeed it was! Reverting that decresed the time spent to 8.44 second.

sylee957 commented on sympy/core/basic.py:
```diff
@@ -619,7 +619,7 @@ def canonical_variables(self):
         return reps
 
-    def rcall(self, *args):
+    def rcall(self, *args, **kwargs):
```
The addition of `kwargs` now serves no purpose in this PR.
I guess it can be either reverted or kept as feature, but if it is going to be added, it needs release note entry.

oscarbenjamin replied:
Why would `rcall` have kwargs?

mcpl-sympy replied:
I think I'll revert it in this PR.

sylee957 commented:
I had been reading *"Functional Differential Geometry", Sussman and Wisdom.*, which contains the philosophy behind the designs of `diffgeom` module,
But I approve that the book doesn't say anything about whether coordinate transformations should be mutable or not, and can only be a matter of implementation details.

I think that the doc directive for `CoordSystemStructure` should be added in `diffgeom.rst`, because it should be one of the core feature of the `diffgeom` module after the API changes.
Since `Introduction` section completely empty there, a basic tutorial can be added to how to define coordinate systems for beginner users, but this can be optional for now.

mcpl-sympy commented:
@sylee957 
Sure thing. I will add right away.

mcpl-sympy commented:
@sylee957
I think #19451 might as well be implemented and documented in this PR.

JSS95 committed:
Resolve entangled conflict with master

mcpl-sympy commented:
Since `Str` is introduced in master branch, coordinate transform relation can now be included in `CoordSystem.args`.  
This makes most of previous commits unnecessary, so I decided to clean up the history and force push here.

Another significant change: `CoordinateSymbol` class is introduced to replace `CoordSystem._dummies`.

Tons of more changes are still needed, and test needs to be properly modified.

JSS95 committed:
Remove import of deleted object

JSS95 committed:
Fix repr printing of CoordSystem

Modified docstrings as well

JSS95 committed:
Fix CoordinateSymbol

JSS95 committed:
Fix imports in rn.py

JSS95 committed:
Fix docstring of Point.py

JSS95 committed:
Enhance latex printing of Manifold and Point

JSS95 committed:
Enhance latex printing of Patch and CoordSystem

JSS95 committed:
Add properties to BaseScalarField

JSS95 committed:
Fix deprecated printings

JSS95 committed:
Modify docstring of BaseVectorField

JSS95 committed:
Fix constructions and docstrings of remaining classes

JSS95 committed:
Fix docstring of functions

JSS95 committed:
Fix doctests

JSS95 committed:
Modify test_class_structure

JSS95 committed:
Fix test_diffgeom.py

JSS95 committed:
Fix other tests

JSS95 committed:
Add CoordinateSymbol to test_args

JSS95 committed:
Delete diffgeom classes from test_pickling

Diffgeom's classes cannot be properly pickled
without completely removing deprecated features.

JSS95 committed:
Improve code quality

JSS95 committed:
Fix flake8 failure

JSS95 committed:
Remove unnecessary tests

Since classes in diffgeom are all Basic now, these
tests are already guaranteed by core's tests.

JSS95 committed:
Fix printing tests

JSS95 committed:
Add CoordinateSymbol to diffgeom.rst

JSS95 committed:
Fix References underline length

mcpl-sympy commented:
@sylee957 
Please review the commit.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -86,25 +114,34 @@ class Patch(Atom):
-    # Contains a reference to the parent manifold in order to be able to access
-    # other patches.
-    def __new__(cls, name, manifold):
+    def __new__(cls, name, manifold, **kwargs):
```
Does this need any keyword arguments?

sylee957 replied:
Okay, you may have kept in mind that some sympy keywords like `evaluate` can propagate down

mcpl-sympy replied:
You got it right.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -120,134 +157,155 @@ def dim(self):
+                [str(n) for n in symbols]
+            )
+        obj.patch.coord_systems.append(obj) # deprecated
+        obj._dummies = _deprecated_list(
```
These are private methods, so it can be overwhelming to raise deprecation warnings for `_names` and `_dummies`

mcpl-sympy replied:
I guess you're right. I un-deprecated them.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -321,85 +514,133 @@ def coord_tuple_transform_to(self, to_sys, coords):
 
     def jacobian(self, to_sys, coords):
         """Return the jacobian matrix of a transformation."""
+        SymPyDeprecationWarning(
```
Although I acknowledge that the term `jacobian` was not used consistently across sympy as well as the general scientific community, I'm not sure this had been necessary to deprecate

mcpl-sympy replied:
I'm convinced. Instead, I made `jacobian_matrix` as an alias of `jacobian`.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -1136,31 +1445,44 @@ class CovarDerivativeOp(Expr):
+    def wrt(self):
+        return self.args[0]
+
+    def christoffel(self):
```
Should this be property?

mcpl-sympy replied:
Oops. My mistake.

sylee957 commented on sympy/diffgeom/diffgeom.py:
```diff
@@ -321,85 +514,133 @@ def coord_tuple_transform_to(self, to_sys, coords):
+    index : integer
+
+    Examples
+    ========
```
I'm not sure it documents about the usage of creating coordinate systems without symbols
```
m = Manifold('M', 2)
p = Patch('P', m)
C = CoordSystem('C', p)
C.symbols
```

And now I get some extraneous warning when I create coordinate systems without symbols

```
Class signature 'names' of CoordSystem has been deprecated since SymPy
1.7. Use class signature 'symbols' instead. See
https://github.com/sympy/sympy/issues/19321 for more info.

  deprecated_since_version="1.7"
```
is it intended feature to deprecate such?

mcpl-sympy replied:
This wasn't my intention. I fixed it. Thanks!

sylee957 commented on sympy/printing/latex.py:
```diff
@@ -2506,31 +2506,52 @@ def _print_MatrixHomomorphism(self, h):
-            coords.name, coords.patch.name, coords.patch.manifold.name
+    def _print_Point(self, point):
+        return '{%s}_{%s}' % (
+            self._print(point.coordinates()), self._print(point._coord_sys)
```
I think that `_print_Point` should not be defined because it conflicts with printing of `Point` of `sympy.geometry` or `sympy.vector`. Maybe it was a bad idea to have duplicate names for classes at the first place, but it works better if you just leave it undefined and have some reasonable tuple notation of args.

mcpl-sympy replied:
How about giving `diffgeom.Point` a new name, and make `'Point'` in diffgeom module as an aliasing?

sylee957 replied:
I have thought as a same idea to rename the Point in diffgeom. But it may be noted in the release notes.

mcpl-sympy replied:
Deleted this. I think we can give different name to class and introduce aliasing, but that would better be handled in another PR.

JSS95 committed:
Resolve reviews from sylee957

CoordSystem._dummies is un-deprecated.
CoordSystem.jacobian is un-deprecated.
CovarDerivativeOp.christoffel is fixed to be property.
Generating CoordSystem without symbols names is supported properly.

JSS95 committed:
Un-deprecate CoordSystem._names

JSS95 committed:
Add test supporting CoordSystem without symbol names

JSS95 committed:
Delete _print_Point in latex.py

sachin-4099 commented:
I will request to build only the latest commit. It leads to long queues on TRAVIS.

sylee957 commented:
I've cancelled some old builds

mcpl-sympy commented:
I'm sorry for consecutive pushes.
